iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -917,26 +981,24 @@ static void decideFusableLinalgOps(FuncOp funcOp) {
       // Start with a root operation and fuse its producers.
       Operation *op = linalgOp.getOperation();
       if (!isRootOp(op)) continue;
-      if (op->getAttrOfType<ArrayAttr>(kFusionGroupsAttr)) continue;
-      unsigned currGroupNum = numRootOps++;
-      op->setAttr(kRootOpAttr, builder.getI64IntegerAttr(currGroupNum));
-      for (auto operand : linalgOp.getShapedOperands()) {
-        auto producer = operand.getDefiningOp<linalg::LinalgOp>();
+
+      // This might be a root op that is already fused into another root op.
+      // In that case, it has fusion groups attached on it.
+      SmallVector<int64_t, 1> fusionGroups = getFunsionGroups(op);
+      if (fusionGroups.empty()) {
+        // Otherwise, create a new group.
+        unsigned newGroup = numRootOps++;
+        fusionGroups.push_back(newGroup);
+        op->setAttr(kRootOpAttr, builder.getI64IntegerAttr(newGroup));
+      }
+
+      for (OpOperand &operand : linalgOp.getShapedOpOperands()) {
```

Comment:
So we need to make a distinction between case where elementwise operation is the root operation. We should only be pulling in the producer if the producer has only one use. If the producer has multiple uses its better to leave the producer in its own dispatch region. Think of a matmul which is used in two elementwise operations. You dont want to fuse it with both (the computation is replicated).

Reply:
Also add a test for this kind of pattern.

Reply:
Yup. We only fuse the producer when it just has one user. This is done via `isProducerFusable`, which has the check `hasOneUse`. There is already a test for it, `dont_fuse_conv2d_with_multiple_uses` in `disaptch_linalg_on_tensors.mlir`.